### PR TITLE
Add jsdoc comments for migrate

### PR DESCRIPTION
**Pull request by dylanonelson**
**What kind of change does this PR introduce?**
JSdocs for migrate

**Did you add tests for your changes?**
No

**If relevant, did you update the documentation?**
n/a

**Summary**
Resolves issue #219

**Does this PR introduce a breaking change?**
No

**Other information**
Did my best to describe everything correctly, but I may have made mistakes. I noticed the existing jsdoc comments were a bit inconsistent, so I went with what made sense to me.

### CODE REVIEW

**Comment by ev1stensberg** [comment]
Nice, thanks! Can you run the jsdocs command and commit?

**Comment by fokusferit** [comment]
@dylanonelson thanks for the work! As this PR is quite huge (due to generation of docs), I'd like to invest at least some time to go through the comments :)

And thanks to your PR, I think for the future doc generation should be part of CI not of each PR.

**Comment by ev1stensberg** [comment]
Yup, +1

**Comment by dylanonelson** [comment]
OK sounds good.

File: lib/migrate/bannerPlugin/bannerPlugin.js [hunk 1]
```diff
@@ -1,5 +1,15 @@
 const utils = require("../../utils/ast-utils");
 
+/**
+ *
+ * Transform for BannerPlugin arguments. Consolidates first and second argument (if
+ * both are present) into single options object.
+ *
+ * @param {object} j — jscodeshift top-level import
```
Comment (ev1stensberg) [inline lib/migrate/bannerPlugin/bannerPlugin.js]: `Object`

File: lib/migrate/bannerPlugin/bannerPlugin.js [hunk 2]
```diff
@@ -1,5 +1,15 @@
...
+ * @param {Collection} ast — jscodeshift Collection to transform
+ * @returns {Collection} ast — jscodeshift Collection
```
Comment (ev1stensberg) [inline lib/migrate/bannerPlugin/bannerPlugin.js]: `Node`

File: lib/migrate/extractTextPlugin/extractTextPlugin.js [hunk 3]
```diff
@@ -1,5 +1,15 @@
 const utils = require("../../utils/ast-utils");
 
+/**
+ *
+ * Check whether `node` is the invocation of the plugin denoted by `pluginName`
+ *
+ * @param {object} j — jscodeshift top-level import
+ * @param {Node} node — ast node to check
+ * @param {string} pluginName
+ * @returns {boolean}
```
Comment (ev1stensberg) [inline lib/migrate/extractTextPlugin/extractTextPlugin.js]: `Boolean`, `Object`, `String`, could you add `pluginName - name of the plugin`?

File: lib/migrate/extractTextPlugin/extractTextPlugin.js [hunk 4]
```diff
@@ -10,6 +20,15 @@ function findInvocation(j, node, pluginName) {
 	return found;
 }
 
+/**
+ *
+ * Transform for ExtractTextPlugin arguments. Consolidates arguments into single options object.
+ *
+ * @param {object} j — jscodeshift top-level import
```
Comment (ev1stensberg) [inline lib/migrate/extractTextPlugin/extractTextPlugin.js]: `Object`, `Node` for ast

File: lib/migrate/index.js [hunk 5]
```diff
@@ -48,7 +48,7 @@ function transformSingleAST(ast, source, transformFunction) {
  * @param { String } source - Source file contents
  * @param { Array<Function> } transformations - List of trnasformation functions in defined the
  * order to apply. By default all defined transfomations.
- * @param { Object } options - Reacst formatting options
+ * @param { Object } options - Recast formatting options
```
Comment (ev1stensberg) [inline lib/migrate/index.js]: Could you add this JSdoc too?

File: lib/migrate/loaderOptionsPlugin/loaderOptionsPlugin.js [hunk 6]
```diff
@@ -4,6 +4,16 @@ const createOrUpdatePluginByName = require("../../utils/ast-utils")
 	.createOrUpdatePluginByName;
 const safeTraverse = require("../../utils/ast-utils").safeTraverse;
 
+/**
+ *
+ * Transform which adds context information for old loaders by injecting a `LoaderOptionsPlugin`
+ *
+ * @param {object} j — jscodeshift top-level import
```
Comment (ev1stensberg) [inline lib/migrate/loaderOptionsPlugin/loaderOptionsPlugin.js]: `Object`

File: lib/migrate/loaderOptionsPlugin/loaderOptionsPlugin.js [hunk 7]
```diff
@@ -4,6 +4,16 @@ const createOrUpdatePluginByName = require("../../utils/ast-utils")
...
+ * @param {Collection} ast — jscodeshift Collection to transform
+ * @returns {Collection} ast — jscodeshift Collection
```
Comment (ev1stensberg) [inline lib/migrate/loaderOptionsPlugin/loaderOptionsPlugin.js]: `Node`

File: lib/migrate/loaders/loaders.js [hunk 8]
```diff
@@ -1,5 +1,16 @@
...
 
+/**
+ *
+ * Transform for loaders. Transforms pre- and postLoaders into enforce options,
+ * moves loader configuration into rules array, transforms query strings and
+ * props into loader options, and adds -loader suffix to loader names.
+ *
+ * @param {object} j — jscodeshift top-level import
+ * @param {Collection} ast — jscodeshift Collection to transform
+ * @returns {Collection} ast — jscodeshift Collection
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: `Node` and `Object`

File: lib/migrate/loaders/loaders.js [hunk 9]
```diff
@@ -36,8 +47,8 @@ module.exports = function(j, ast) {
 	 *     loader: 'css'
 	 *   }]
 	 * }
-	 * @param  {Node} path Must be an ObjectExpression
-	 * @return {Node}      []
+	 * @param  {NodePath} path — wrapped ast object expression
+	 * @returns {NodePath} path
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: Use `Node` for for everything ast related

File: lib/migrate/loaders/loaders.js [hunk 10]
```diff
@@ -83,6 +94,14 @@ module.exports = function(j, ast) {
 		return path;
 	};
 
+	/**
+	 *
+	 * Puts query parameters from loader value into options object
+	 *
+	 * @param {NodePath} p — wrapped ast object expression node for loader object
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: `Node`

File: lib/migrate/loaders/loaders.js [hunk 11]
```diff
@@ -106,6 +125,14 @@ module.exports = function(j, ast) {
 		return j.objectExpression([loaderProp, queryProp]);
 	};
 
+	/**
+	 *
+	 * Determine whether a loader has a query string
+	 *
+	 * @param {NodePath} p — wrapped ast object expression node for loader object
+	 * @returns {boolean} Whether the loader value contains a query string
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: `Node` and `Boolean`

File: lib/migrate/loaders/loaders.js [hunk 12]
```diff
@@ -117,12 +144,14 @@ module.exports = function(j, ast) {
...
-	 * If the path value is `loaders` and it's located in `module` object
-	 * we assume it's the loader's section
+	 * Check if the identifier is the `loaders` prop in the `module` object.
+	 * If the path value is `loaders` and it’s located in `module` object
+	 * we assume it’s the loader's section.
 	 *
-	 * @param  {Node} path [description]
-	 * @return {Node}      [description]
+	 * @param  {NodePath} path — wrapped ast identifier node
+	 * @returns {boolean}
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: `Boolean`, `Node`

File: lib/migrate/loaders/loaders.js [hunk 13]
```diff
@@ -135,11 +164,12 @@ module.exports = function(j, ast) {
 		]) === "module";
 
 	/**
-	 * Puts node path that is pre- or postLoader into `enforce` key
+	 * Puts pre- or postLoader into `loaders` object and adds the appropriate `enforce` property
 	 *
-	 * @param {Node} p []
-	 * @returns {*}		[]
+	 * @param {NodePath} p — wrapped ast object expression that has a key for either 'preLoaders' or 'postLoaders'
+	 * @returns {NodePath} p
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: `Node`

File: lib/migrate/loaders/loaders.js [hunk 14]
```diff
@@ -171,10 +201,9 @@ module.exports = function(j, ast) {
 	};
 
 	/**
-	 * Find pre and postLoaders
+	 * Find pre- and postLoaders in ast and put them into the `loaders` array
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: Should be "find pre and postLoaders in the ast and put them into the `loaders` array"

File: lib/migrate/loaders/loaders.js [hunk 15]
```diff
@@ -185,9 +214,8 @@ module.exports = function(j, ast) {
 
 	/**
 	 * Convert top level `loaders` to `rules`
-	 * See https://webpack.js.org/configuration/module/#module-rules
-	 * @param  {[type]} IPath [description]
-	 * @return {[type]}       [description]
+	 *
+	 * @returns {Collection} jscodeshift Collection
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: @ returns {Node} ast - JScodeshift node

File: lib/migrate/loaders/loaders.js [hunk 16]
```diff
@@ -197,9 +225,9 @@ module.exports = function(j, ast) {
 			.forEach(p => (p.value.name = "rules"));
 
 	/**
-	 * Converts 'loader' and 'loaders' to Array of {Rule.Use}
+	 * Convert `loader` and `loaders` to Array of {Rule.Use}
 	 *
-	 * @returns {Node} []
+	 * @returns {Collection} jscodeshift Collection
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: @ returns {Node} ast - JScodeshift ast

File: lib/migrate/loaders/loaders.js [hunk 17]
```diff
@@ -219,7 +247,7 @@ module.exports = function(j, ast) {
 			.forEach(createArrayExpressionFromArray);
 
 	/**
-	 * Finds loaders with options encoded as query string and replaces it with options obejct
+	 * Find loaders with options encoded as query string and replace it with options obejct
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: `object` typo

File: lib/migrate/loaders/loaders.js [hunk 18]
```diff
@@ -237,8 +265,7 @@ module.exports = function(j, ast) {
 	 *   }
 	 * }
 	 *
-	 * @type {[type]}
-	 * @returns {Node} []
+	 * @returns {Collection} jscodeshift Collection
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: @ returns {Node} ast - JScodeshift ast

File: lib/migrate/loaders/loaders.js [hunk 19]
```diff
@@ -249,7 +276,7 @@ module.exports = function(j, ast) {
 			.replaceWith(createLoaderWithQuery);
 
 	/**
-	 * Finds nodes with `query` key and replaces it with `options`
+	 * Find nodes with `query` key and replace it with `options`
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: a `query` key, `replaces it`.

File: lib/migrate/loaders/loaders.js [hunk 20]
```diff
@@ -261,7 +288,8 @@ module.exports = function(j, ast) {
 	 * {
 	 *   options: { ... }
 	 * }
-	 * @returns {Node} []
+	 *
+	 * @returns {Collection} jscodeshift Collection
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: @ returns {Node} ast - JScodeshift ast

File: lib/migrate/loaders/loaders.js [hunk 21]
```diff
@@ -271,9 +299,9 @@ module.exports = function(j, ast) {
 			.replaceWith(j.identifier("options"));
 
 	/**
-	 * Adds required `-loader` suffix to loader with missing suffix
-	 * i.e. for `babel` it should generate `babel-loader`
-	 * @returns {Node} []
+	 * Add required `-loader` suffix to loader with missing suffix
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: to a loader*,  @ returns {Node} ast - JScodeshift ast

File: lib/migrate/outputPath/outputPath.js [hunk 22]
```diff
@@ -1,5 +1,14 @@
 const utils = require("../../utils/ast-utils");
 
+/**
+ *
+ * Transform which adds `path.join` call to `output.path` literals
+ *
+ * @param {object} j — jscodeshift top-level import
+ * @param {Collection} ast — jscodeshift Collection to transform
```
Comment (ev1stensberg) [inline lib/migrate/outputPath/outputPath.js]: `Object` and `Node`

File: lib/migrate/removeDeprecatedPlugins/removeDeprecatedPlugins.js [hunk 23]
```diff
@@ -1,6 +1,16 @@
 const chalk = require("chalk");
 const utils = require("../../utils/ast-utils");
 
+/**
+ *
+ * Find deprecated plugins and remove them from the `plugins` array, if possible.
+ * Otherwise, warn the user about removing deprecated plugins manually.
+ *
+ * @param {object} j — jscodeshift top-level import
```
Comment (ev1stensberg) [inline lib/migrate/removeDeprecatedPlugins/removeDeprecatedPlugins.js]: Object and node..

File: lib/migrate/removeJsonLoader/removeJsonLoader.js [hunk 24]
```diff
@@ -1,6 +1,24 @@
 const utils = require("../../utils/ast-utils");
 
+/**
+ *
+ * Transform which removes the json loader from all possible declarations
+ *
+ * @param {object} j — jscodeshift top-level import
+ * @param {Collection} ast — jscodeshift Collection to transform
```
Comment (ev1stensberg) [inline lib/migrate/removeJsonLoader/removeJsonLoader.js]: Object and node

File: lib/migrate/removeJsonLoader/removeJsonLoader.js [hunk 25]
```diff
@@ -1,6 +1,24 @@
...
+ * @returns {Collection} ast — jscodeshift Collection
+ */
+
 module.exports = function(j, ast) {
+	/**
+	 *
+	 * Remove the loader with name `name` from the given NodePath
+	 *
+	 * @param {NodePath} path — NodePath to remove the loader from
+	 * @param {string} name — the name of the loader to remove
```
Comment (ev1stensberg) [inline lib/migrate/removeJsonLoader/removeJsonLoader.js]: String and Node

File: lib/migrate/resolve/resolve.js [hunk 26]
```diff
@@ -1,3 +1,12 @@
+/**
+ *
+ * Transform which consolidates the `resolve.root` configuration option into the `resolve.modules` array
+ *
+ * @param {object} j — jscodeshift top-level import
```
Comment (ev1stensberg) [inline lib/migrate/resolve/resolve.js]: Big O in Object and Node

File: lib/migrate/resolve/resolve.js [hunk 27]
```diff
@@ -16,6 +25,15 @@ module.exports = function transformer(j, ast) {
 		return modules.length > 0 && modules[0];
 	};
 
+	/**
+	 *
+	 * Add a `modules` property to the `resolve` object or update the existing one
+	 * based on what is already in `resolve.root`
+	 *
+	 * @param {NodePath} p — wrapped ast property node that represents the `resolve` property
```
Comment (ev1stensberg) [inline lib/migrate/resolve/resolve.js]: Node and elaboration in returns

File: lib/migrate/uglifyJsPlugin/uglifyJsPlugin.js [hunk 28]
```diff
@@ -1,5 +1,14 @@
 const findPluginsByName = require("../../utils/ast-utils").findPluginsByName;
 
+/**
+ *
+ * Transform which adds a `sourceMap: true` option to instantiations of the UglifyJsPlugin
+ *
+ * @param {object} j — jscodeshift top-level import
```
Comment (ev1stensberg) [inline lib/migrate/uglifyJsPlugin/uglifyJsPlugin.js]: Big O and Node

**Comment by ev1stensberg** [comment]
Planned to have this merged and fixing that myself, but let's just fix it now 🍶

File: lib/migrate/index.js (hunk 5 above)
Comment (dylanonelson) [reply lib/migrate/index.js]: Sure, no problem. Do you mean the jsdoc for the properties of the `options` object?

File: lib/migrate/resolve/resolve.js (hunk 27 above)
Comment (dylanonelson) [reply lib/migrate/resolve/resolve.js]: Not sure what you mean here by “elaboration in returns”?

**Comment by dylanonelson** [comment]
OK I did my best to make all the requested changes and also scanned through all the docs to make sure they matched your description—`{Node}` instead of `{Collection}` and `jscodeshift ast` instead of `jscodeshift Collection`, etc. I left a couple of questions on your comments, the big one being what you meant in `lib/migrate/index.js` about adding another jsdoc on L51.

Thanks for the review!

File: lib/migrate/resolve/resolve.js (hunk 27 above)
Comment (ev1stensberg) [reply lib/migrate/resolve/resolve.js]: Elaborate what p returns a bit more 👍

File: lib/migrate/index.js (hunk 5 above)
Comment (ev1stensberg) [reply lib/migrate/index.js]: No, you forgot to do this/add the extra *. And `Array<Function>` is wrong, should be `Function[]`

**Comment by ev1stensberg** [comment]
Re commented, thanks for checking it out!

File: lib/migrate/resolve/resolve.js (hunk 27 above)
Comment (dylanonelson) [reply lib/migrate/resolve/resolve.js]: I added a description to the `@returns` tag. Do you want me to add descriptions for all of them? In cases where it seemed self-explanatory I omitted the description.

**Comment by ev1stensberg** [comment]
Thanks! Will re-review later today! ♥️

File: lib/migrate/loaders/loaders.js [hunk 29]
```diff
@@ -219,7 +248,7 @@ module.exports = function(j, ast) {
 			.forEach(createArrayExpressionFromArray);
 
 	/**
-	 * Finds loaders with options encoded as query string and replaces it with options obejct
+	 * Find loaders with options encoded as a query string and replace the string with an options object
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: interactive verb? `replaces`

File: lib/migrate/loaders/loaders.js [hunk 30]
```diff
@@ -249,7 +277,7 @@ module.exports = function(j, ast) {
 			.replaceWith(createLoaderWithQuery);
 
 	/**
-	 * Finds nodes with `query` key and replaces it with `options`
+	 * Find nodes with a `query` key and replace it with `options`
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: interactive indication, so: `replaces`

File: lib/migrate/extractTextPlugin/extractTextPlugin.js [hunk 31]
```diff
@@ -1,5 +1,15 @@
 const utils = require("../../utils/ast-utils");
 
+/**
+ *
+ * Check whether `node` is the invocation of the plugin denoted by `pluginName`
+ *
+ * @param {Object} j - jscodeshift top-level import
+ * @param {Node} node - ast node to check
+ * @param {String} pluginName - name of the plugin
+ * @returns {Boolean}
```
Comment (ev1stensberg) [inline lib/migrate/extractTextPlugin/extractTextPlugin.js]: Could you give a bit of description what the boolean indicates in the `@returns` ?

File: lib/migrate/loaders/loaders.js [hunk 32]
```diff
@@ -36,8 +47,9 @@ module.exports = function(j, ast) {
 	 *     loader: 'css'
 	 *   }]
 	 * }
-	 * @param  {Node} path Must be an ObjectExpression
-	 * @return {Node}      []
+	 *
+	 * @param  {Node} path - object expression ast
+	 * @returns {Node} path
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: Could you explain what the returns value does/is?

File: lib/migrate/loaders/loaders.js [hunk 33]
```diff
@@ -117,12 +145,14 @@ module.exports = function(j, ast) {
...
-	 * If the path value is `loaders` and it's located in `module` object
-	 * we assume it's the loader's section
+	 * Check if the identifier is the `loaders` prop in the `module` object.
+	 * If the path value is `loaders` and it’s located in `module` object
+	 * we assume it’s the loader's section.
 	 *
-	 * @param  {Node} path [description]
-	 * @return {Node}      [description]
+	 * @param {Node} path - identifier ast
+	 * @returns {Boolean}
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: Could you elaborate? Thanks 🎧

File: lib/migrate/loaders/loaders.js [hunk 34]
```diff
@@ -135,11 +165,12 @@ module.exports = function(j, ast) {
 		]) === "module";
 
 	/**
-	 * Puts node path that is pre- or postLoader into `enforce` key
+	 * Puts pre- or postLoader into `loaders` object and adds the appropriate `enforce` property
 	 *
-	 * @param {Node} p []
-	 * @returns {*}		[]
+	 * @param {Node} p - object expression ast that has a key for either 'preLoaders' or 'postLoaders'
+	 * @returns {Node} p
```
Comment (ev1stensberg) [inline lib/migrate/loaders/loaders.js]: Would be a 10/10 if there was a small explaining part here as well

File: lib/migrate/loaders/loaders.js (hunk 29 above)
Comment (dylanonelson) [reply lib/migrate/loaders/loaders.js]: I'm a little confused about this. You want to change `replace` to `replaces` but leave `Find` in the same sentence? It seems like they should match. I skimmed through some of the other comments and it looks like sometimes the doclets use the `replace` version and sometimes they use the `replaces` version. If you want, I can make them all consistent. Or there may be something special about this one case that I don't understand—just let me know.

Comment (ev1stensberg) [reply lib/migrate/loaders/loaders.js]: its a small pick, so no worries 👍

**Comment by ev1stensberg** [comment]
@ematipico and @fokusferit can you re-review? Thanks bois!

**Pull request merged**